pipeline.py

```diff
@@ -72,31 +92,85 @@ class TextProcessingPipeline:
-        pass
+        articles = self.corpus_manager.get_articles().values()
+        for article in articles:
+            raw_text = article.get_raw_text()
+            processed_tokens = self._process(raw_text)
+
+            cleaned_tokens = []
+            single_tagged_tokens = []
+            multiple_tagged_tokens = []
+
+            for processed_token in processed_tokens:
+                cleaned_tokens.append(processed_token.get_cleaned())
+                single_tagged_tokens.append(processed_token.get_single_tagged())
+                multiple_tagged_tokens.append(processed_token.get_multiple_tagged())
+
+            article.save_as(' '.join(cleaned_tokens), ArtifactType.cleaned)
+            article.save_as(' '.join(single_tagged_tokens), ArtifactType.single_tagged)
+            article.save_as(' '.join(multiple_tagged_tokens), ArtifactType.multiple_tagged)
 
     def _process(self, raw_text: str):
         """
         Processes each token and creates MorphToken class instance
         """
-        pass
+        cleaned_text = raw_text.replace('-\n', '')
+        plain_text_analysis = Mystem().analyze(cleaned_text)
+        morph_analyzer = pymorphy2.MorphAnalyzer()
+
+        tokens = []
+        for single_word_analysis in plain_text_analysis:
+            if not single_word_analysis.get('analysis') or not single_word_analysis.get('text'):
+                continue
+            morphological_token = MorphologicalToken(original_word=single_word_analysis['text'])
+            morphological_token.normalized_form = single_word_analysis['analysis'][0]['lex']
+            morphological_token.tags_mystem = single_word_analysis['analysis'][0]['gr']
+            morphological_token.tags_pymorphy = morph_analyzer.parse(single_word_analysis['text'])[0].tag
+            tokens.append(morphological_token)
+        return tokens
 
 
 def validate_dataset(path_to_validate):
     """
     Validates folder with assets
     """
-    pass
+    path = Path(path_to_validate)
+    if not path.exists():
+        raise FileNotFoundError
+    if not path.is_dir():
+        raise NotADirectoryError
+    if not list(path.iterdir()):
+        raise EmptyDirectoryError
+    raws = list(path.glob('*_raw.txt'))
+    metas = list(path.glob('*_meta.json'))
+    all_files = list(path.iterdir())
+    for file in all_files:
+        if '_raw.txt' in file.name:
+            with open(file, 'r', encoding='utf-8') as text_file:
+                text = text_file.read()
+                if not text:
+                    raise InconsistentDatasetError
+    if not len(metas) == len(raws):
+        raise InconsistentDatasetError
+    raw_indices = sorted(list(map(lambda x: int(x.name.split('_')[0]), raws)))
+    meta_indices = sorted(list(map(lambda x: int(x.name.split('_')[0]), metas)))
+    if raw_indices[0] != 1 or meta_indices[0] != 1:
+        raise InconsistentDatasetError
+    if not raw_indices == meta_indices:
+        raise InconsistentDatasetError
 
```

Comment:
you are also expected to check that file numeration does not have any slips 
for example, imagine a folder with files with following ids: 1, 2, 3, 5, 6, 7, 8, 9
it is not acceptable as the file with id 4 is missing
will your validation system catch such a case?

Reply:
Yes, I have rewritten the code. Now, it does the following:
1) I introduce 2 variables - "raw_txt" and "meta_json". They are both equal to 0.
2) then I iterate through all files files in the folder
if file format is "*_raw.txt" or "*_meta.json" variables increase by 1, and loop continues; otherwise, there is an exception InconsistentDatasetError.
The code is on lines 167-179